contracts/core/mailbox/src/execute.rs

```diff
@@ -164,17 +117,33 @@ pub fn dispatch(
             .to_msg(MAILBOX_VERSION, nonce, config.local_domain, &info.sender)?;
     let msg_id = msg.id();
 
-    let (required_hook_value, hook_value) = get_required_value(
-        deps.as_ref(),
-        &info,
-        required_hook.as_str(),
+    let base_gas = hook::quote_dispatch(
+        &deps.querier,
+        dispatch_msg.get_hook_addr(deps.api, default_hook)?,
+        dispatch_msg.metadata.clone().unwrap_or_default(),
+        msg.clone(),
+    )?
+    .gas_amount;
+
+    let required_gas = hook::quote_dispatch(
+        &deps.querier,
+        &required_hook,
         dispatch_msg.metadata.clone().unwrap_or_default(),
-        msg.clone().into(),
+        msg.clone(),
+    )?
+    .gas_amount;
+
+    // assert gas received is satisfies required gas
+    let mut total_gas = required_gas.clone().into_iter().try_fold(
+        Coins::try_from(base_gas.clone())?,
+        |mut acc, gas| {
+            acc.add(gas)?;
+            StdResult::Ok(acc)
+        },
     )?;
```

Comment:
if the denominations are not the same, we should return both rather than add them together

Reply:
`cosmwasm_std::Coins`'s add method contains the [logic](https://github.com/CosmWasm/cosmwasm/blob/8fe095fa75fe8e8bb1fa75a030d9cce2dfd7a391/packages/std/src/coins.rs#L173) that you mentioned.